objective.py

```diff
@@ -20,12 +23,21 @@ def set_data(self, X, y):
         self.n_features = self.X.shape[1]
 
     def compute(self, beta):
+        # compute residuals
         if self.fit_intercept:
             beta, intercept = beta[:self.n_features], beta[self.n_features:]
         diff = self.y - self.X.dot(beta)
         if self.fit_intercept:
             diff -= intercept
-        return .5 * diff.dot(diff) + self.lmbd * abs(beta).sum()
+        # compute primal objective and duality gap
+        p_obj = .5 * diff.dot(diff) + self.lmbd * abs(beta).sum()
+        theta = diff / self.lmbd
+        theta /= norm(self.X.T @ theta, ord=np.inf)
+        d_obj = (norm(self.y) ** 2 / 2. - self.lmbd ** 2 *
+                 norm(self.y / self.lmbd - theta) ** 2 / 2)
+        return dict(value=p_obj,
+                    nnz=(beta != 0).sum(),
```

Comment:
I would use more explicit names as these are the ones in the plots.

so `sparsity` and `dual_gap`?